src/Altinn.Apps/AppTemplates/AspNet/Altinn.App.PlatformServices/Implementation/ProcessEngine.cs

```diff
@@ -0,0 +1,130 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Threading.Tasks;
+using Altinn.App.Common.Process;
+using Altinn.App.Core.Interface;
+using Altinn.App.Core.Models;
+using Altinn.App.PlatformServices.Models;
+using Altinn.App.Services.Helpers;
+using Altinn.App.Services.Interface;
+
+namespace Altinn.App.Core.Implementation
+{
+    /// <summary>
+    /// The process engine is responsible for all BMPN related functionality
+    ///
+    /// It will call processChange handler that is responsible
+    /// for the business logic happening for any process change.
+    /// </summary>
+    public class ProcessEngine : IProcessEngine
+    {
+        private readonly IProcessChangeHandler _processChangeHandler;
+
+        private readonly ProcessHelper processHelper;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ProcessEngine"/> class.
+        /// </summary>
+        public ProcessEngine(
+                IProcessChangeHandler processChangeHandler,
+                IProcess processService)
+        {
+            _processChangeHandler = processChangeHandler;
+            using Stream bpmnStream = processService.GetProcessDefinition();
+            processHelper = new ProcessHelper(bpmnStream);
+        }
+
+        /// <summary>
+        /// Move process to next element in process
+        /// </summary>
+        public async Task<ProcessChangeContext> Next(ProcessChangeContext processChange)
+        {
+            string currentElementId = processChange.Instance.Process.CurrentTask?.ElementId;
+
+            if (currentElementId == null)
+            {
+                processChange.ProcessMessages = new System.Collections.Generic.List<ProcessChangeInfo>();
+                processChange.ProcessMessages.Add(new ProcessChangeInfo() { Message = $"Instance does not have current task information!", Type = "Conflict" });
+                return processChange;
+            }
+
+            if (currentElementId.Equals(processChange.RequestedProcessElementId))
+            {
+                processChange.ProcessMessages = new System.Collections.Generic.List<ProcessChangeInfo>();
+                processChange.ProcessMessages.Add(new ProcessChangeInfo() { Message = $"Requested process element {processChange.RequestedProcessElementId} is same as instance's current task. Cannot change process.", Type = "Conflict" });
+                return processChange;
```

Comment:
Here we return a ProcessChangeContext object where the property FailedProcessChange is unchanged (false?). This means that methods like for example ProcessController.NextElement will assume success and return Ok. Completely discarding the ProcessMessage that indicate conflict

Reply:
Good catch. Fixed